### DISCUSSION TITLE
RCS in Fossify Messages

### DISCUSSION THREAD

**Discussion by DeadSecTr** [FossifyOrg/General-Discussion#490] [Ideas]
Is RCS possible?

**Comment by inson1** [comment]
https://github.com/SimpleMobileTools/Simple-SMS-Messenger/issues/157

**Reply by yavko** [reply]
Maybe the code in https://android.googlesource.com/platform/packages/services/Telephony/+/refs/heads/main/testapps/TestRcsApp/ can be observed

**Reply by naveensingh** [reply]
I saw that too, may be worth looking into as there are no open-source players in the RCS game, and when RCS becomes more popular, people will have no choice but to use Google Messages. 

One caveat is the internet permission.

**Reply by foolishgrunt** [reply]
Oh man, I would support the _heck_ out of this! A legitimate FOSS RCS client is literally #1 on my wish list, especially considering that even Apple is adopting it later this year. I would gladly grant internet permission to Fossify if it could make it happen.

To be fair, I know "support" from a non-programmer is mostly cheap talk. But I could build code, test code, donate an old phone, donate money... anything to help make it happen!

**Reply by inson1** [reply]
Also this was mentioned in SMT https://github.com/android-rcs/rcsjta. Did somebody try it?

**Reply by foolishgrunt** [reply]
^That's a very stale repository - I imagine it wouldn't still be useful.

https://github.com/Hirohumi/rust-rcs-client seems to be a bit more current, but I couldn't tell you the first thing about how complete (or useful) it is.

**Comment by webdingspowah** [comment]
Here's all the documentation from GSMA from 04 June 2024 for RCS Universal Profile 2.7 (latest): 
https://www.gsma.com/solutions-and-impact/technologies/networks/wp-content/uploads/2024/06/RCC.07-v14.0.pdf
https://www.gsma.com/solutions-and-impact/technologies/networks/wp-content/uploads/2024/06/RCC.11-v12.0.pdf
https://www.gsma.com/solutions-and-impact/technologies/networks/wp-content/uploads/2024/06/RCC.14-v10.0.pdf
https://www.gsma.com/solutions-and-impact/technologies/networks/wp-content/uploads/2024/06/RCC.15-v8.0.pdf
https://www.gsma.com/solutions-and-impact/technologies/networks/wp-content/uploads/2024/06/RCC.71-v2.7.pdf

Here are the old versions in case they are needed:
1.0:
https://www.gsma.com/solutions-and-impact/technologies/networks/wp-content/uploads/2017/07/RCC.71_v1.0.pdf
2.0:
https://www.gsma.com/solutions-and-impact/technologies/networks/wp-content/uploads/2017/07/RCC.71_v2.0.pdf
2.1:
https://www.gsma.com/solutions-and-impact/technologies/networks/wp-content/uploads/2017/09/RCC.71_v2.1.pdf
2.2:
https://www.gsma.com/solutions-and-impact/technologies/networks/wp-content/uploads/2018/05/Universal-Profile-RCC.71-v2.2.pdf
2.3:
https://www.gsma.com/solutions-and-impact/technologies/networks/wp-content/uploads/2018/12/RCC.71-v2.3.pdf
2.4:
https://www.gsma.com/solutions-and-impact/technologies/networks/wp-content/uploads/2019/10/RCC.71-v2.4.pdf
2.5:
https://www.gsma.com/solutions-and-impact/technologies/networks/wp-content/uploads/2024/07/RCC.71-v2.5.pdf
2.6:
https://www.gsma.com/solutions-and-impact/technologies/networks/wp-content/uploads/2024/07/RCC.71-v2.6.pdf

**Reply by webdingspowah** [reply]
Here's all the documentation from GSMA from 28 February 2025 for RCS Universal Profile 3.0 with MLS E2EE (latest):
https://www.gsma.com/solutions-and-impact/technologies/networks/wp-content/uploads/2025/03/RCC.07-v15.0.pdf
https://www.gsma.com/solutions-and-impact/technologies/networks/wp-content/uploads/2025/03/RCC.11-v13.0.pdf
https://www.gsma.com/solutions-and-impact/technologies/networks/wp-content/uploads/2025/03/RCC.14-v11.0.pdf
https://www.gsma.com/solutions-and-impact/technologies/networks/wp-content/uploads/2025/03/RCC.16-v1.0.pdf
https://www.gsma.com/solutions-and-impact/technologies/networks/wp-content/uploads/2025/03/RCC.71-v3.0.pdf

If you're looking for the MLS document specifically, it's RCC.16.

Available open source implementations of MLS:
https://github.com/openmls/openmls
https://github.com/cisco/mlspp
https://github.com/awslabs/mls-rs
https://gitlab.matrix.org/matrix-org/mls-ts/

**Reply by webdingspowah** [reply]
Here's all the documentation from GSMA from 18 July 2025 for RCS Universal Profile 3.1 with MLS E2EE (latest):
https://www.gsma.com/solutions-and-impact/technologies/networks/wp-content/uploads/2012/10/RCC.07-v16.0.pdf
https://www.gsma.com/solutions-and-impact/technologies/networks/wp-content/uploads/2012/10/RCC.14-v12.0.pdf
https://www.gsma.com/solutions-and-impact/technologies/networks/wp-content/uploads/2012/10/RCC.16-v2.0.pdf
https://www.gsma.com/solutions-and-impact/technologies/networks/wp-content/uploads/2012/10/RCC.71-v3.1.pdf

**Comment by webdingspowah** [comment]
This client claims to support RCS: https://github.com/vidit135g/Replify-Messenger

**Reply by foolishgrunt** [reply]
I rate that claim "unlikely."

There was a single release over four years ago, with no code activity since then. I think it's more likely that the dev forked QKSMS and wrote a "wishlist" (including RCS support) into the README, but never got around to implementing it. And I'm certainly not about to install an unknown app just to test it. :)

**Reply by KishkaCat** [reply]
^  Even if it does work, I highly doubt it would meet modern RCS E2EE standards.

**Reply by beedell-roke** [reply]
> https://github.com/FossifyOrg/Messages/discussions/34#discussioncomment-12634025

@Shad0wL3g3nd, is *standard* RCS encryption really E2E? That might be vendor-specific. [^1]

[^1]: [`reddit.com/r/UniversalProfile/comments/xinnha/comment/ip3ynot`](https://www.reddit.com/r/UniversalProfile/comments/xinnha/comment/ip3ynot/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1#:~:text=E2ee%20are%20using%20Google%20messages..%20the%20others%20are%20using%20Samsung%20messages.)

**Reply by KishkaCat** [reply]
Google implies there is E2EE as they reference iPhones _NOT_ having it, therefore RCS does support it.

![Screenshot_20250326-181829.png](https://github.com/user-attachments/assets/df1c58f6-9722-4867-b299-94f5ca0e805c)
[https://support.google.com/messages?p=rcs_chats_faq](https://support.google.com/messages?p=rcs_chats_faq)

**Reply by naveensingh** [reply]
https://en.wikipedia.org/wiki/Rich_Communication_Services#Encryption_support

**Reply by KishkaCat** [reply]
Okay, if you expand it, it says: "Google Messages offers end-to-end encryption for your RCS messages with other Google Messages users. The GSM Association (GSMA) sets Mobile and RCS standards and we are currently working to define security standards cross platform." So it at least offers E2EE between Android Devices using Google Messages. The source may be using old code that doesn't support it either way.

**Reply by beedell-roke** [reply]
> https://github.com/FossifyOrg/Messages/discussions/34#discussioncomment-12634166

Thanks, @naveensingh. It appears that it was previously vendor-specific, but is now standard: [^1]

> [End-to-end (E2E) encryption](https://en.wikipedia.org/wiki/End-to-end_encryption) via [MLS](https://en.wikipedia.org/wiki/Messaging_Layer_Security) is specified by Universal Profile RCS 3.0 as of March 2025, but has not been implemented yet. Previously, E2E encryption was not a feature of RCS specified by GSMA, instead deferring to the individual clients to establish E2E encryption. In September 2024, the GSMA announced it was working on bringing interoperable E2E encryption to the Universal Profile RCS standard.

[^1]: [`en.wikipedia.org/w/index.php?title=Rich_Communication_Services&oldid=1280793231#cite_ref-:24_11-1`](https://en.wikipedia.org/w/index.php?title=Rich_Communication_Services&oldid=1280793231#cite_ref-:24_11-1:~:text=End%2Dto%2Dend%20(E2E)%20encryption%20via%20MLS,the%20Universal%20Profile%20RCS%20standard.%5B89%5D)

**Comment by ZethiaZ** [comment]
A maintained RCS compatible messenger would be killer, I'd donate for that.

**Reply by KishkaCat** [reply]
Honestly, set up a gofundme or something.

**Comment by KishkaCat** [comment]
Could Google's RBM(RCS Business Messages) be messed with a little to support this? It looks fairly similar, just more of an automotive version of such, so it could be reverse engineered possibly to work? Honestly, take this with a grain of salt.

`https://developers.google.com/business-communications/rcs-business-messaging/guides/get-started/how-it-works`

**Reply by KishkaCat** [reply]
Or actually, even better: ✨source code✨

https://android.googlesource.com/platform/packages/apps/Messaging/

**Reply by atikattar1104** [reply]
@KitCatLove Bro Really Said "Dump The Theory, Take The Implementation." ☠️

Jokes Apart, This Is Great, Only If There Is No Vendor Lock-in, And E2EE Is Maintainable As Well. 

Also, They Seem To Abandon This Code For Some Reason. Check Out Readme From It's Main Branch.

**Reply by KishkaCat** [reply]
> @KitCatLove Bro Really Said "Dump The Theory, Take The Implementation." ☠️
> 
> Jokes Apart, This Is Great, Only If There Is No Vendor Lock-in, And E2EE Is Maintainable As Well. 
> 
> Also, They Seem To Abandon This Code For Some Reason. Check Out Readme From It's Main Branch. 

Good point, but it would at least get base encryption. Even if it's out dated

**Reply by naveensingh** [reply]
> [android.googlesource.com/platform/packages/apps/Messaging](https://android.googlesource.com/platform/packages/apps/Messaging/)

Guys, it's just the AOSP messaging app. There's no encryption there.

**Reply by beedell-roke** [reply]
@naveensingh, ~~I don't see why the OS matters, and Universal Profile RCS 3.0 appears to include cryptographic recommendations.~~

**Reply by beedell-roke** [reply]
Oh, you meant literally `android.googlesource.com/platform/packages/apps/Messaging`. Apologies! 😆

**Comment by Hoerli1337** [comment]
So if I have understood everything here correctly, is RCS support not yet being implemented or what is the status?
It would be really great if we had that in the app :)

**Reply by beedell-roke** [reply]
> https://github.com/FossifyOrg/Messages/discussions/34#discussioncomment-13236457

@Hoerli1337, it's not being developed (yet).

**Reply by Hoerli1337** [reply]
ooh okay :(

**Reply by KishkaCat** [reply]
Unfortunately RCS isn't an open thing to be implemented. It appears that Google, Samsung, and Apple are keeping it from smaller projects(Like Fossify).

**Reply by beedell-roke** [reply]
> https://github.com/FossifyOrg/Messages/discussions/34#discussioncomment-13250829

@KitCatLove, from what I see in relevant GSMA documentation, much of it is available for implementation. Does anything specific prevent all RCS support, rather than prevent 100% conformance?

**Reply by KishkaCat** [reply]
Can you clarify which RCS support Documentation you see, for I can only find RCS for business documentation.

**Reply by beedell-roke** [reply]
> https://github.com/FossifyOrg/Messages/discussions/34#discussioncomment-13250911

@KitCatLove, I was polling my own memory when I said that, but I don't seem to be able to locate a citation anymore. Maybe I was thinking of the AOSP API commits visible in Gerrit. 🤷

**Reply by KishkaCat** [reply]
That's Fine; No worries!

**Reply by foolishgrunt** [reply]
Are these not they? 

https://github.com/FossifyOrg/Messages/discussions/34#discussioncomment-11378386

**Reply by beedell-roke** [reply]
> https://github.com/FossifyOrg/Messages/discussions/34#discussioncomment-13252955

@foolishgrunt, that was my first estimate too, but I had a look at a few, and they don't appear particularly technical.

**Reply by KishkaCat** [reply]
@foolishgrunt Those are also from late 2022, and RCS has changed since then.

**Reply by beedell-roke** [reply]
> https://github.com/FossifyOrg/Messages/discussions/34#discussioncomment-13255787

@KitCatLove, and no replacements exist?

**Reply by KishkaCat** [reply]
Not that I see... I feel like replacing it with the Signal Protocol would work well, but that is not compatible with RCS :/

**Reply by beedell-roke** [reply]
> https://github.com/FossifyOrg/Messages/discussions/34#discussioncomment-13256720

@KitCatLove, myriad alternatives exist that are vastly superior. Matrix, XMPP, Signal, etcetera. The sole reason that RCS matters, like its SMS and MMS predecessors, is that carriers treat SMS, MMS, and RCS packets differently to all else in their payment plans. I have no idea of why.

**Reply by beedell-roke** [reply]
> https://github.com/FossifyOrg/Messages/discussions/34#discussioncomment-13256739

Ah, did you think that I was referring to the protocol? If so, I was instead referring to updated versions of the cited documents.

**Reply by KishkaCat** [reply]
Oh yes, my apologies. I can do some more research.

**Reply by KishkaCat** [reply]
@RokeJulianLockhart I just dumped the Google Messages app; I'll look through it; and you can have a copy if you want.

Dump
> https://wormhole.app/LO3BL9#dXFPN5ZdhA_eK79I0YWIqg

**Reply by rhysperry111** [reply]
Might not be the best idea providing a sketchy download link and also associating an open-source project with dubiously-legal dumps of proprietary applications.

**Reply by beedell-roke** [reply]
> associating an open-source project with dubiously-legal dumps of proprietary applications

@rhysperry111, it looks like an extracted `application/vnd.android.package-archive`. I don't think extracting `application/zip` contents [^1] can be considered to be decompilation, and there's a slight EU exemption for interoperability: [^2]

> In the European Union, reverse engineering is legal as long as it is for interoperability purposes, whatever the license may say on the subject. Reverse engineering does not give you the right to publish your findings, though.

We're not publishing papers on the internal structure of copyrighted software. We're just seeing how an application communicates with a server.

[^1]: [`stackoverflow.com/revisions/15934887/1`](https://stackoverflow.com/revisions/15934887/1#content:~:text=Any%20APK%20file%20is%20a%20valid%20ZIP%20file.)

[^2]: [`security.stackexchange.com/revisions/30375/1`](https://security.stackexchange.com/revisions/30375/1#:~:text=In%20the%20USA%2C%20it%20is,to%20publish%20your%20findings%2C%20though.)

> Might not be the best idea providing a sketchy download link

There's [`virustotal.com/gui/url-analysis/u-dfa315433638fc72b0efa2c9b286cdfb76c300c260a12d65b42257a11bd4f680-1748377905`](https://www.virustotal.com/gui/url-analysis/u-dfa315433638fc72b0efa2c9b286cdfb76c300c260a12d65b42257a11bd4f680-1748377905#:~:text=https://wormhole.app/download%2Dstream/mrk3ajZ4TB2qgwXDxqElGg):

![image](https://github.com/user-attachments/assets/398fef16-1745-4c49-9fd4-c89f00b487d1)

**Reply by beedell-roke** [reply]
> https://github.com/FossifyOrg/Messages/discussions/34#discussioncomment-13288000

@KitCatLove, thanks! How did you dump it? Maybe, I can nab a VM with AOSP 16 Beta + GMS, and dump that for the latest stuff.

**Reply by rhysperry111** [reply]
> @rhysperry111, it looks like an extracted application/vnd.android.package-archive. I don't think extracting application/zip contents can be considered to be decompilation, and there's a slight EU exemption for interoperability:

I absolutely agree with you from a technical standpoint, however it's the sort of thing that has been known in the past to poke large corps in just the wrong way and is usually just best not bringing up especially in public channels.

> We're not publishing papers on the internal structure of copyrighted software. We're just seeing how an application communicates with a server.

Again, technically I do agree, but I don't have the money to fight Big Tech™'s lawyers :P

> There's [virustotal.com/gui/url-analysis/u-dfa315433638fc72b0efa2c9b286cdfb76c300c260a12d65b42257a11bd4f680-1748377905](https://www.virustotal.com/gui/url-analysis/u-dfa315433638fc72b0efa2c9b286cdfb76c300c260a12d65b42257a11bd4f680-1748377905#:~:text=https://wormhole.app/download%2Dstream/mrk3ajZ4TB2qgwXDxqElGg):

That link you posted shows VirusTotal got a 404 from the page so it likely was a broken link or was getting blocked somehow. Generally it's not even worth trying to work out whether weird links are actually sketchy or not as there a plenty of non-sketchy providers that could've been used and this still does nothing to verify that there isn't something nasty about the download itself.

**Reply by KishkaCat** [reply]
> > [#34 (reply in thread)](https://github.com/FossifyOrg/Messages/discussions/34#discussioncomment-13288000)
> 
> @KitCatLove, thanks! How did you dump it? Maybe, I can nab a VM with AOSP 16 Beta + GMS, and dump that for the latest stuff.

I [downloaded the app from the Google Play Store](https://play.google.com/store/apps/details?id=com.google.android.apps.messaging), then used [Nova Launcher](https://play.google.com/store/apps/details?id=com.teslacoilsw.launcher) to get the APK(you can also use [Files by Google](https://play.google.com/store/apps/details?id=com.google.android.apps.nbu.files) by clicking apps and than finding Google Messages and saving it's APK(There are endless video on YouTube)), then I clicked "Open with" on the APK and clicked 'Explore' under the app: [App Manager](https://f-droid.org/en/packages/io.github.muntashirakon.AppManager/). From there on, I long pressed on one of the files/folders, then select all. Next I clicked the more options button, than Copy. I deselected them by clicking the small close button on the actions popup, then clicked the "/" in the location bar to navigate to root. From here, I navigated to "storage/emulated/0/documents". I clicked new in the bottom right, selected "Folder", naming it "Google Messages Dump". Navigate to your new folder. In the top right, click the snowman menu button, then paste. I waited for it to finish, closed the app, opened my file manager app, navigated into documents, found the folder, long pressed it, then clicked compress, and sent it to my computer which I uploaded here.

**Reply by KishkaCat** [reply]
> Might not be the best idea providing a sketchy download link and also associating an open-source project with dubiously-legal dumps of proprietary applications.

This website is the only one I trust for sharing files.

**Reply by KishkaCat** [reply]
> > Might not be the best idea providing a sketchy download link
> 
> There's [`virustotal.com/gui/url-analysis/u-dfa315433638fc72b0efa2c9b286cdfb76c300c260a12d65b42257a11bd4f680-1748377905`](https://www.virustotal.com/gui/url-analysis/u-dfa315433638fc72b0efa2c9b286cdfb76c300c260a12d65b42257a11bd4f680-1748377905#:~:text=https://wormhole.app/download%2Dstream/mrk3ajZ4TB2qgwXDxqElGg):

That's of the website, not the file...

**Reply by beedell-roke** [reply]
> That link you posted shows VirusTotal got a 404 from the page so it likely was a broken link or was getting blocked somehow.

@rhysperry111, missed that, embarrassingly. Apologies. I'd *presumed* they'd leave the URI available for at least 3 minutes, but they seem to close it on client disconnect. That's quite aggressive culling on the provider's part. I pity anyone with a poor internet connection forced to use it. Irrespective, today's when I learn VT has a low maximum file size limit:

> The ZIP file you are uploading exceeds the size limit (3MB).

**Reply by beedell-roke** [reply]
> That's of the website, not the file...

@KitCatLove, I'd supplied the URI affiliated with the `Content-Disposition: attachment`.

**Reply by beedell-roke** [reply]
> https://github.com/FossifyOrg/Messages/discussions/34#discussioncomment-13288754

@KitCatLove, I wouldn't quite call that a "dump", so if that's all, good luck gleaning anything from Bionic-compiled binaries! How do you intend to debug it? The farthest I've ever forayed there is LLDB.

**Reply by KishkaCat** [reply]
My appologies,

**Reply by KishkaCat** [reply]
I found a new method I will try tommorow.

**Reply by KishkaCat** [reply]
I got actual source this time. @RokeJulianLockhart Do you have a private method of contact? Ex. Signal?

**Reply by beedell-roke** [reply]
> https://github.com/FossifyOrg/Messages/discussions/34#discussioncomment-13299462

@KitCatLove, thanks for notifying me! I've an account for everything (although [`@rokejulianlockhart:matrix.org`](https://matrix.to/#/@rokejulianlockhart:matrix.org) or [`discord.com/users/638098902570237965`](https://discord.com/users/638098902570237965) would be better).

However, better yet, why not file a new [`github.com/RokeJulianLockhart/RokeJulianLockhart/discussions`](https://github.com/RokeJulianLockhart/RokeJulianLockhart/discussions/new/choose)? That is, if you don't mind it being public. I've no fear that Google shall ever care, so I don't mind, and it might mean others with some competence can have a look.

**Reply by KishkaCat** [reply]
I have Matrix, I just don't want this project taken down...

**Reply by beedell-roke** [reply]
@KitCatLove, nothing we do in our own repositories could impact Fossify, otherwise removals of downstream forks would impact upstreams, which would be nonsensical.

#### Retroactive Modification

See [`RokeJulianLockhart/RokeJulianLockhart/discussions/8`](https://github.com/RokeJulianLockhart/RokeJulianLockhart/discussions/8#discussion-8385759).

**Reply by KishkaCat** [reply]
Quite true.

**Reply by KishkaCat** [reply]
@RokeJulianLockhart I created a discussion.

**Reply by webdingspowah** [reply]
@RokeJulianLockhart @KitCatLove
Did you even read the documentation I posted? That's the specification for RCS. What do you mean it's not technical? It instructs you how to implement, to my understanding, the entirety of RCS over hundreds of pages of extremely detailed documentation. It even has schemas you can directly copy. Those 5 documents are for version 2.7, released in 2024. No, it's not from 2022. Why did you even think it's from 2022? It literally says June 2024 on the very first page of all 5 documents. Did you even open them?

Also, I just posted links to version 3.0, including the MLS E2EE specification as well. Plus, a list of all available open source MLS implementations.

**Reply by KishkaCat** [reply]
> @RokeJulianLockhart @KitCatLove
> Did you even read the documentation I posted? That's the specification for RCS. What do you mean it's not technical? It instructs you how to implement, to my understanding, the entirety of RCS over hundreds of pages of extremely detailed documentation. It even has schemas you can directly copy. Those 5 documents are for version 2.7, released in 2024. No, it's not from 2022. Why did you even think it's from 2022? It literally says June 2024 on the very first page of all 5 documents. Did you even open them?
> 
> Also, I just posted links to version 3.0, including the MLS E2EE specification as well. Plus, a list of all available open source MLS implementations.

No. I don't recall seeing it to be honest?

**Reply by beedell-roke** [reply]
> https://github.com/FossifyOrg/Messages/discussions/34#discussioncomment-13399172

@webdingspowah, I tried to, but it was incomprehensible to a layman like myself. If you look at my past comments here, I explicitly asked the thread's participants whether your documentation explained the E2EE specification, but was informed by someone that it didn't anymore due to its age. I don't consider myself to be an authority on the matter, so I believed them.

Thank you for the new documentation.

**Reply by webdingspowah** [reply]
Well, good luck. Because whoever is brave enough to take on this monumental task is going to need it. The new document for the E2EE specification alone is another 90 pages of information that will need to be read and implemented, on top of all the other documentation that outlines the specification. Not even Google and Apple have this completed yet. At the time of writing this message, nobody in the world is on RCS version 3.0. At the moment, Apple is on version 2.4, and Google is on 2.6 with partial support for 2.7. Which means Google is two and a half years behind, and Apple is nearly six years behind.

**Reply by KishkaCat** [reply]
> > @RokeJulianLockhart @KitCatLove
> > Did you even read the documentation I posted? That's the specification for RCS. What do you mean it's not technical? It instructs you how to implement, to my understanding, the entirety of RCS over hundreds of pages of extremely detailed documentation. It even has schemas you can directly copy. Those 5 documents are for version 2.7, released in 2024. No, it's not from 2022. Why did you even think it's from 2022? It literally says June 2024 on the very first page of all 5 documents. Did you even open them?
> > Also, I just posted links to version 3.0, including the MLS E2EE specification as well. Plus, a list of all available open source MLS implementations.
> 
> No. I don't recall seeing it to be honest?

I do remember it looking back; sorry...

**Reply by bonswouar** [reply]
FYI apparently it might be integrated into /e/ at some point: https://github.com/octoshrimpy/quik/issues/124#issuecomment-2955242262

**Comment by arneal1** [comment]
I'm new to Fossify and would love to use it as my primary app, but without the ability to react to texts, it is lacking.

**Reply by foolishgrunt** [reply]
I'm right there with you, but let's not make it sounds like Naveen is slacking off.

Currently, there is no usable RCS client for Android other than Google Messages - it is not technically feasible for any FOSS client to implement at this point.

**Reply by KishkaCat** [reply]
I think Samsung's messaging app has RCS compatability as well.

**Reply by beedell-roke** [reply]
@KitCatLove, [`substack.com/profile/139839157-henry/note/c-146388452`][1] explains why:

> Samsung Messages is a special case, apparently using Google Messages’ internals.

[1]: https://substack.com/profile/139839157-henry/note/c-146388452#:~:text=Samsung%20Messages%20is%20a%20special%20case%2C%20apparently%20using%20Google%20Messages%E2%80%99%20internals.

**Comment by foolishgrunt** [comment]
This will be an interesting thread to follow. RCS as part of microG? Yes, please!

https://github.com/microg/GmsCore/issues/2994

EDIT: I misread the request. The bounty is for microG to allow Google Messages to use RCS, not for a broad-based implementation of RCS within microG that other apps could connect to.

**Reply by Random90** [reply]
Given the direction Google is taking android, they will block any attempt at implementing RCS outside google messages. Sideloading will be thing of the past soon too. It was fun while it lasted.

**Reply by naveensingh** [reply]
> Sideloading will be thing of the past soon too. It was fun while it lasted.

Hey, at least they were considerate enough to [allow ADB installs](https://developer.android.com/developer-verification/guides/faq#general).

**Reply by Random90** [reply]
> Hey, at least they were considerate enough to [allow ADB installs](https://developer.android.com/developer-verification/guides/faq#general).

For now at least, but not for long I suppose

**Reply by atikattar1104** [reply]
They Are Restricted To Official Androids. Stuff Like Lineage OS, Etc. Are Safe. Also, Most Phones Are Now Compatible With Project Treble. You Can Sideload A GSI Image From Any Of Your Own Favourite AOSP (Un)Maintained Fork, And Enjoy Android Restriction Free.

**Reply by beedell-roke** [reply]
@atikattar1104, how does that relate to RCS?
